flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamTwoInputSelectableProcessor.java

```diff
@@ -0,0 +1,433 @@
+		.select(2)
+		.build()
+		.getInputMask();
+
+	private static final int INPUT1_ID = 1;
+	private static final int INPUT2_ID = 2;
+
+	// ---------------- Metrics ------------------
+
+	private final WatermarkGauge input1WatermarkGauge;
+	private final WatermarkGauge input2WatermarkGauge;
+
+	private Counter numRecordsIn;
+
+	public StreamTwoInputSelectableProcessor(
+		Collection<InputGate> inputGates1,
+		Collection<InputGate> inputGates2,
+		TypeSerializer<IN1> inputSerializer1,
+		TypeSerializer<IN2> inputSerializer2,
+		Object lock,
+		IOManager ioManager,
+		StreamStatusMaintainer streamStatusMaintainer,
+		TwoInputStreamOperator<IN1, IN2, ?> streamOperator,
+		WatermarkGauge input1WatermarkGauge,
+		WatermarkGauge input2WatermarkGauge) {
+
+		checkState(streamOperator instanceof InputSelectable);
+
+		// create a NetworkInput instance for each input
+		StreamTwoInputStreamStatusHandler streamStatusHandler = new StreamTwoInputStreamStatusHandler(
+			streamStatusMaintainer, lock);
+		this.input1 = new NetworkInput(INPUT1_ID, inputGates1, inputSerializer1, streamStatusHandler, ioManager);
+		this.input2 = new NetworkInput(INPUT2_ID, inputGates2, inputSerializer2, streamStatusHandler, ioManager);
+
+		this.lock = checkNotNull(lock);
+
+		this.streamOperator = checkNotNull(streamOperator);
+		this.inputSelector = (InputSelectable) streamOperator;
+
+		this.input1WatermarkGauge = input1WatermarkGauge;
+		this.input2WatermarkGauge = input2WatermarkGauge;
+
+		this.auxiliaryHandler = new AuxiliaryHandler();
+
+		this.listenFutures = new CompletableFuture[]{null, null};
+		this.isFinished = new boolean[]{false, false};
+	}
+
+	/**
+	 * Notes that it must be called after calling all operator's open(). This ensures that
+	 * the first input selection determined by the operator at and before opening is effective.
+	 */
+	public void init() {
+		inputSelection = inputSelector.nextSelection();
+
+		availableInputsMask.set(TWO_INPUT_ANY_MASK);
+
+		lastReadingInputMask = (int) InputSelection.SECOND.getInputMask();
+
+		if (numRecordsIn == null) {
+			try {
+				numRecordsIn = ((OperatorMetricGroup) streamOperator
+					.getMetricGroup()).getIOMetricGroup().getNumRecordsInCounter();
+			} catch (Exception e) {
+				LOG.warn("An exception occurred during the metrics setup.", e);
+				numRecordsIn = new SimpleCounter();
+			}
+		}
+	}
+
+	public boolean processInput() throws Exception {
+		// cache inputs reference on the stack, to make the code more JIT friendly
+		final NetworkInput input1 = this.input1;
+		final NetworkInput input2 = this.input2;
+
+		while (continuousProcessing) {
+			final int selectionMask = (int) inputSelection.getInputMask();
+			final int availableMask = availableInputsMask.get();
+
+			int readingInputMask = selectionMask & availableMask;
+			if (readingInputMask == TWO_INPUT_ANY_MASK) {
+				// the input selection is `ALL` and both inputs are available, we read two inputs in turn
+				readingInputMask -= lastReadingInputMask;
```

Comment:
That's a very good structure. But for performance reasons, the code here optimizes the fair selection algorithm in the case of two inputs. That is, the algorithm implemented here satisfies only two inputs. If the encapsulated  `fairSelectNextIndex()` is added to the `InputSelection` class, I understand that the algorithm should be universal and adapt to any number of inputs, then the algorithm is relatively complex and poor efficiency for two inputs.

Reply:
Yes, it would be more general. Whether it would be less efficient, I don't know but I would bet that not necessarily. In the end, this logic should be encapsulated and unit tested one way or another, either:

1. propper `fairSelectNextIndex` implementation
2. or implement currently specialised version `fairSelectNextIndexOutOf2` which would throw `UnsupportedOperationException` for input selections > 2

Reply:
For the moment, the second one, +1